merkletree/merkletree.go

```diff
@@ -0,0 +1,288 @@
+// merkletree provides tools for calculating the Merkle root of a dataset, for
+// creating a proof that a piece of data is in a Merkle tree of a given root,
+// and for verifying proofs that a piece of data is in a Merkle tree of a given
+// root.
+package merkletree
+
+import (
+	"bytes"
+	"hash"
+)
+
+// A Tree takes data as leaves and returns the merkle root. Each call to 'Push'
+// adds one leaf to the merkle tree. Calling 'Root' returns the Merkle root.
+// The Tree also constructs proof that a single leaf is a part of the tree. The
+// leaf can be chosen with 'SetIndex'.
+type Tree struct {
```

Comment:
I guess the other issue with naming it merkletree is that `tree := merkletree.Tree` kinda stutters. But I don't expect people to be working with the `Tree` type directly anyway.